include/highfive/bits/H5Reference_misc.hpp

```diff
@@ -0,0 +1,99 @@
+/*
+ *  Copyright (c), 2020, EPFL - Blue Brain Project
+ *
+ *  Distributed under the Boost Software License, Version 1.0.
+ *    (See accompanying file LICENSE_1_0.txt or copy at
+ *          http://www.boost.org/LICENSE_1_0.txt)
+ *
+ */
+
+#ifndef H5REFERENCE_MISC_HPP
+#define H5REFERENCE_MISC_HPP
+
+#include <string>
+
+namespace HighFive {
+
+inline Reference::Reference(const Object& location, const Object& object)
+    : parent_id(location.getId()) {
+
+    const size_t maxLength = 255;
+    char buffer[maxLength + 1];
+    if (H5Iget_name(object.getId(), buffer, maxLength) <= 0) {
+        HDF5ErrMapper::ToException<DataTypeException>("Invalid object or location");
+    }
+    obj_name = std::string(buffer);
+}
+
+
+void Reference::create_ref(hobj_ref_t* refptr) const {
+    if (H5Rcreate(refptr, parent_id, obj_name.c_str(), H5R_OBJECT, -1) < 0) {
+        HDF5ErrMapper::ToException<ReferenceException>(
+            std::string("Unable to create the reference for \"") + obj_name +
+            "\":");
+    }
+}
+
+ObjectType Reference::getType(const Object& location) const {
+    hid_t res;
+    if ((res = H5Rdereference(location.getId(), H5R_OBJECT, &href)) < 0) {
+        HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
+    }
+    return Object(res).getType();
+}
+
+template <typename T>
+T Reference::dereference(const Object& location) const {
+    static_assert(std::is_base_of<Object, T>::value != 0,
+        "Can only return a subtype of HighFive::Object");
+    hid_t res;
+    if ((res = H5Rdereference(location.getId(), H5R_OBJECT, &href)) < 0) {
+        HDF5ErrMapper::ToException<ReferenceException>("Unable to dereference.");
+    }
+    auto obj = Object(res);
+    T h5_obj(obj);
+    return h5_obj;
+}
+
+namespace details {
+
+template <>
+struct data_converter<std::vector<Reference>> {
```

Comment:
Hopefully by inheriting from `template<Container> data_converter` you can drop a lot of code here. (rebased in the latest master?)